### PR TITLE
show code coverage with Codecov.io

### PR DESCRIPTION
**Pull request by wilzbach**
Hi all, the idea for this automation is simple, let's ensure (1) that no PR decreases the code coverage, (2) help reviewers judging code coverage improvements PRS, (3) provide simple tools for contributors to see which parts are uncovered.

You can preview this [here](https://codecov.io/gh/wilzbach/phobos/tree/5fc9eb90076101c0266fb3491ac68527d3520fba/std).

That being said, this won't be perfect in the near future as Travis can only run Linux and Mac OS builds, so special versioned paths will not be treated right. The idea here is to get an easy 90% solution that already helps reviewers in most cases (=a lot better than 0%). In the future we might add Codecov reporting to the AutoTester.

Now to this PR:

1) Thanks to the help from @stevepeak, Codecov.io is able to recognize the code coverage output from our builds
2) I haven't figured out yet how I can get the test runner to report the test coverage, for now I settled with
`make -f posix.mak std.test etc.test`
3) I ran into a couple of issues during my first attempt, so don't merge yet. At the moment I was lazy and just commented them out, but I/we should definitely investigate the root cause of them:
- `entropy` in std.numeric (and thus schwarzSort in std.algorithm.sorting)
- `utf-16le` in std.encoding
-  `iso-8859-1 not recognized` for std.net.curl
4) As the newest `druntime` and `dmd` are needed, we now have to move `phobos` in its own folder and then clone the other repos. It makes the Travis script a bit ugly, better ideas are of course welcome.

(if everything goes well, CodeCov should comment on this PR)

### CODE REVIEW

File: std/numeric.d [hunk 1]
```diff
@@ -1946,7 +1947,7 @@ if (isInputRange!Range &&
     return result;
 }
 
-@safe unittest
+unittest
```
Comment (WalterBright) [inline std/numeric.d]: why turn off @safe?

File: std/numeric.d [hunk 2]
```diff
@@ -1925,6 +1925,7 @@ ElementType!Range entropy(Range)(Range r) if (isInputRange!Range)
     Unqual!(typeof(return)) result = 0.0;
     foreach (e; r)
     {
+        writeln(e);
```
Comment (WalterBright) [inline std/numeric.d]: why?

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -2231,73 +2231,73 @@ unittest
...
-    arr[1] = lowEnt;
-    arr[2] = highEnt;
-
-    schwartzSort!(entropy, q{a < b})(arr);
-    assert(arr[0] == lowEnt);
-    assert(arr[1] == midEnt);
-    assert(arr[2] == highEnt);
-    assert(isSorted!("a < b")(map!(entropy)(arr)));
-}
+//unittest
```
Comment (WalterBright) [inline std/algorithm/sorting.d]: too comment out code use:
```
version (none) { ... }
```
not commenting out each line.

File: posix.mak [hunk 4]
```diff
@@ -358,7 +358,7 @@ unittest/%.run : $(ROOT)/unittest/test_runner
 # The mktemp business is needed so .o files don't clash in concurrent unittesting.
 %.test : %.d $(LIB)
 	T=`mktemp -d /tmp/.dmd-run-test.XXXXXX` && \
-	  $(DMD) -od$$T $(DFLAGS) -main -unittest $(LIB) -defaultlib= -debuglib= $(LINKDL) -cov -run $< && \
+	  $(DMD) -od$$T $(DFLAGS) -g -main -unittest $(LIB) -defaultlib= -debuglib= $(LINKDL) -cov -run $< && \
```
Comment (WalterBright) [inline posix.mak]: no need for `-g`

File: std/numeric.d (hunk 2 above)
Comment (wilzbach) [inline std/numeric.d]: As said I experienced weird errors with this. I haven't got the time to look deeper in them, but the error from Travis is reproducible with `make - f posix.mak std.numeric.test`.

File: posix.mak (hunk 4 above)
Comment (wilzbach) [inline posix.mak]: Without `-g` I got gibberish stack traces on Travis, but yeah as soon as the other problems are resolved, I can remove that one ;-)

**Comment by stevepeak** [comment]
@wilzbach Nice job :) Let me know if you have any questions about Codecov configuration.

File: .travis.yml [hunk 5]
```diff
@@ -2,23 +2,35 @@ sudo: false
...
  # enforce all-man style
- - grep -nE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$'  $(find . -name '*.d'); test $? -eq 1
+ - (cd phobos && grep -nE '(if|for|foreach|foreach_reverse|while|unittest|switch|else|version) .*{$'  $(find . -name '*.d'); test $? -eq 1)
  # at the moment libdparse has problems to parse some modules (->excludes)
- - ./dsc --config .dscanner.ini --styleCheck $(find etc std -type f -name '*.d' | grep -vE 'std/traits.d|std/typecons.d|std/conv.d') -I.
+ - (cd phobos && ../Dscanner/dsc --config ../.dscanner.ini --styleCheck $(find etc std -type f -name '*.d' | grep -vE 'std/traits.d|std/typecons.d|std/conv.d') -I.)
+ # test code coverage
+ - (cd dmd && make -f posix.mak)
+ - (cd druntime && make -f posix.mak)
+ - (cd phobos && make -f posix.mak $(find std etc -name "*.d" | sed "s/[.]d$/.test/" | grep -vE '(std.encoding|net.curl)' )
```
Comment (wilzbach) [inline .travis.yml]: Encoding's aren't registered properly for single tests - see e.g. https://github.com/dlang/phobos/pull/4493 for more infos

File: std/numeric.d [hunk 6]
```diff
@@ -1923,10 +1923,10 @@ result is greater than or equal to $(D max).
 ElementType!Range entropy(Range)(Range r) if (isInputRange!Range)
 {
     Unqual!(typeof(return)) result = 0.0;
-    foreach (e; r)
+    for (;!r.empty; r.popFront)
```
Comment (wilzbach) [inline std/numeric.d]: See https://issues.dlang.org/show_bug.cgi?id=16227

**Comment by wilzbach** [comment]
Travis is passing finally :tada:, so this experiment is ready.

As mentioned earlier there are some weird bootstrapping issues with `std.encoding`, but they are [reported](https://issues.dlang.org/show_bug.cgi?id=16291) and being worked on. Thus these modules are excluded currently from the coverage report.

You can preview the automatic [coverage report here](https://codecov.io/gh/dlang/phobos/compare/f9dd0f8e3f965fbcc59b34bd9d666a9f9fc9fe93...d8c181b4b506d3af29620e86bd576c3d82b9df6b/tree/std).

In case you are unfamiliar with `CodeCov`, it's a service that processes coverage reports and will hopefully help us to spot coverage regression and judge coverage improvements.

The behavior of `CodeCov` is [highly configurable](https://github.com/codecov/support/wiki/codecov.yml), so if anything annoys or will annoy you, we most likely will find a solution ;-)

**Pull request merged**